django/db/models/deletion.py

```diff
@@ -31,6 +39,12 @@ def PROTECT(collector, field, sub_objs, using):
     )
 
 
+def RESTRICT(collector, field, sub_objs, using):
+    collector.add_restricted_objects(field, sub_objs)
+    if not field.null:
```

Comment:
I don't think that we should check `field.null`, dependency should be added also for nullable fields, e.g. `test_restrict_generic_foreign_key_and_no_fast_delete` fails after the following change 
```diff
diff --git a/tests/delete/models.py b/tests/delete/models.py                                                                                           
index ecf52fe8ff..28cd707843 100644                                                                                                                    
--- a/tests/delete/models.py
+++ b/tests/delete/models.py
@@ -192,7 +192,7 @@ class GenericB2(models.Model):
 
 
 class GenericDeleteBottom(models.Model):
-    generic_b1 = models.ForeignKey(GenericB1, models.RESTRICT)
+    generic_b1 = models.ForeignKey(GenericB1, models.RESTRICT, null=True)
     content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
     object_id = models.PositiveIntegerField()
     generic_b2 = GenericForeignKey()
```

Reply:
You're right, the dependency exists even for nullable fields because we're not calling `add_field_update()` here. Fixed in https://github.com/django/django/pull/7364/commits/b1a586f64ee1f1332cc2718d9075d9d81b319e9a.